http/restful-ws/src/main/java/io/cloudevents/http/restful/ws/CloudEventsProvider.java

```diff
@@ -0,0 +1,149 @@
+    }
+
+    private byte[] bufferBodyInput(InputStream inputStream) throws IOException {
+        if (inputStream == null) {
+            return null;
+        }
+
+        ByteArrayOutputStream buffer = new ByteArrayOutputStream();
+
+        int nRead;
+        byte[] data = new byte[1024];
+        while ((nRead = inputStream.read(data, 0, data.length)) != -1) {
+            buffer.write(data, 0, nRead);
+        }
+
+        buffer.flush();
+        return buffer.toByteArray();
+    }
+
+    @Override
+    public boolean isWriteable(Class<?> type, Type genericType, Annotation[] annotations, MediaType mediaType) {
+        return CloudEvent.class.isAssignableFrom(type);
+    }
+
+    @Override
+    public void writeTo(CloudEvent event, Class<?> type, Type genericType, Annotation[] annotations, MediaType mediaType, MultivaluedMap<String, Object> httpHeaders, OutputStream entityStream) throws IOException, WebApplicationException {
+        Optional<String> structuredEncodingFormat = Arrays
+            .stream(annotations)
+            .filter(a -> a.annotationType().equals(StructuredEncoding.class))
+            .map(a -> ((StructuredEncoding) a).value())
+            .findFirst();
+
+        if (structuredEncodingFormat.isPresent()) {
+            writeStructured(
+                event,
+                structuredEncodingFormat.get(),
+                new RestfulWSMessageVisitor(httpHeaders, entityStream)
+            );
+        } else {
+            writeBinary(
+                event,
+                new RestfulWSMessageVisitor(httpHeaders, entityStream)
+            );
+        }
+    }
+
+    private <V extends MessageVisitor<V, Void> & BinaryMessageVisitor<Void>> void writeBinary(CloudEvent input, V visitor) {
+        input.asBinaryMessage().visit(visitor);
+    }
+
+    private <V extends MessageVisitor<V, Void> & BinaryMessageVisitor<Void>> void writeStructured(CloudEvent input, EventFormat format, V visitor) {
+        input
+            .asStructuredMessage(format)
+            .visit(visitor);
+    }
+
+    private <V extends MessageVisitor<V, Void> & BinaryMessageVisitor<Void>> void writeStructured(CloudEvent input, String formatString, V visitor) {
+        EventFormat format = EventFormatProvider.getInstance().resolveFormat(formatString);
+
+        if (format == null) {
+            throw new IllegalArgumentException("Cannot resolve format " + formatString);
+        }
+
+        input
+            .asStructuredMessage(format)
+            .visit(visitor);
+    }
+
+    @Override
+    public void filter(ClientRequestContext requestContext) throws IOException {
+        if (Utils.isCloudEventEntity(requestContext.getEntity())) {
+            EventFormat format = EventFormatProvider.getInstance().resolveFormat(requestContext.getMediaType().toString());
+
+            if (format != null) {
+                writeStructured(
```

Comment:
Uh, so you write in the filter? That's weird, but OK.

Reply:
It is, but it works

Reply:
With `MessageBodyWriter` some corner cases aren't working